PMM-6395 VMagent integration

f41gh7 opened the pull request:
https://github.com/Percona-Lab/pmm-submodules/pull/1185
https://github.com/percona/pmm/pull/603
https://github.com/percona/pmm-admin/pull/122
https://github.com/percona/pmm-agent/pull/201
https://github.com/percona/pmm-managed/pull/501
https://github.com/percona/grafana-dashboards/pull/1034
https://github.com/Percona-Lab/pmm-api-tests/pull/140

yareach committed:
[PMM-6756] Added go 1.15.x. (#120)

* [PMM-6756] Added go 1.15.x.

* PMM-6756 Update

Co-authored-by: Andrii Skomorokhov <andrii.skomorokhov@percona.com>

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/signed)](https://cla-assistant.percona.com:443/percona/pmm-admin?pullRequest=122) <br/>All committers have signed the CLA.

BupycHuk commented on commands/inventory/add_agent_mongodb_exporter.go:
```diff
@@ -101,4 +103,5 @@ func init() {
 	AddAgentMongodbExporterC.Flag("skip-connection-check", "Skip connection check").BoolVar(&AddAgentMongodbExporter.SkipConnectionCheck)
 	AddAgentMongodbExporterC.Flag("tls", "Use TLS to connect to the database").BoolVar(&AddAgentMongodbExporter.TLS)
 	AddAgentMongodbExporterC.Flag("tls-skip-verify", "Skip TLS certificates validation").BoolVar(&AddAgentMongodbExporter.TLSSkipVerify)
+	AddAgentMongodbExporterC.Flag("use_vmagent", "When false, pull model is used. By default, its true").Default("true").BoolVar(&AddAgentMongodbExporter.UseVMAgent)
```
Have you discussed about usage of this flag?

f41gh7 replied:
@AlekSi  suggested to add `use_vmagent` to inventory api, with it user can disable push model for some exporters. 

By default, push model detects at the server by matching client version for management api and by reading this flag for inventory api.

BupycHuk replied:
does this mean that vm-agent doesn't support pull model?

f41gh7 replied:
old code

BupycHuk commented on commands/inventory/add_agent_mongodb_exporter.go:
```diff
@@ -101,4 +103,5 @@ func init() {
 	AddAgentMongodbExporterC.Flag("skip-connection-check", "Skip connection check").BoolVar(&AddAgentMongodbExporter.SkipConnectionCheck)
 	AddAgentMongodbExporterC.Flag("tls", "Use TLS to connect to the database").BoolVar(&AddAgentMongodbExporter.TLS)
 	AddAgentMongodbExporterC.Flag("tls-skip-verify", "Skip TLS certificates validation").BoolVar(&AddAgentMongodbExporter.TLSSkipVerify)
+	AddAgentMongodbExporterC.Flag("use_vmagent", "When false, pull model is used. By default, its true").Default("true").BoolVar(&AddAgentMongodbExporter.UseVMAgent)
```
The second problem that I see here, it's related to true by default https://github.com/alecthomas/kingpin/issues/243.
There is nothing in help how to disable it.

f41gh7 replied:
old code

f41gh7 committed:
PMM-6395 - adds metrics-mode flow,
changes inventory api with push-metrics mode,
changes management api with metrics-mode enum auto, push and pull,
adds setup flag for node register metrics-mode for node-exporter

f41gh7 committed:
adds enum for metrics-mode command

f41gh7 committed:
apply format

AlekSi committed:
Merge branch 'master' into PMM-6395-vmagent-integration

f41gh7 committed:
Adds docker image for tests

f41gh7 committed:
Merge branch 'PMM-6395-vmagent-integration' of github.com:percona/pmm-admin into pmm-6395

f41gh7 committed:
reverts docker-compose, its not needed

BupycHuk commented on commands/management/add_mongodb.go:
```diff
@@ -20,6 +20,8 @@ import (
 	"strings"
 
+	"github.com/AlekSi/pointer"
+
```
let's remove this blank line

BupycHuk replied:
and in other places, please.
Only 3 groups in imports
1. go packages
2. packages from other repositories
3. packages from the current repo.

f41gh7 replied:
thanks, will watch for it.

BupycHuk commented on commands/management/management.go:
```diff
@@ -37,4 +37,6 @@ var (
 
 	allServiceTypesKeys = []string{"mysql", "mongodb", "postgresql", "proxysql", "external"}
+
+	metricsModes = []string{"AUTO", "PUSH", "PULL"}
```
let's use lowercase

f41gh7 replied:
updated

BupycHuk commented on commands/config.go:
```diff
@@ -148,4 +153,6 @@ func init() {
 
 	ConfigC.Flag("force", "Remove Node with that name with all dependent Services and Agents if one exist").BoolVar(&Config.Force)
+	ConfigC.Flag("metrics-mode", "Metrics flow mode for agents node-exporter, can be push - agent will push metrics,"+
+		" pull - server scrape metrics from agent  or auto - chosen by server.").EnumVar(&Config.MetricsMode, "AUTO", "PUSH", "PULL")
```
let's use lowercase

BupycHuk replied:
we should set `auto` as default value

BupycHuk replied:
the same in management commands

f41gh7 replied:
done

f41gh7 committed:
updates api

f41gh7 committed:
updates pmm api

f41gh7 committed:
updates commands - lower case for metrics-mode,
fixes imports

f41gh7 committed:
adds default auto for metrics-mode

BupycHuk reviewed (changes requested):
We forgot about proxysql

f41gh7 committed:
updates api, adds proxysql and rds metrics mode

f41gh7 committed:
updates vendor

f41gh7 committed:
adds list agents metrics mode PMM-6399
pmm vendor update

f41gh7 committed:
Adds external exporter push metrics

f41gh7 committed:
fixes enum

f41gh7 committed:
PMM-6399 adds summary command with vmagent targets

f41gh7 committed:
fix confict

f41gh7 committed:
applied filter for list properly

AlekSi committed:
PMM-6395 Update PMM

BupycHuk committed:
Merge branch 'master' into PMM-6395-vmagent-integration
